src/norm2est.cc

```diff
@@ -0,0 +1,207 @@
+// the terms of the BSD 3-Clause license. See the accompanying LICENSE file.
+
+#include "slate/slate.hh"
+#include "internal/internal.hh"
+#include "internal/internal_util.hh"
+#include "slate/internal/mpi.hh"
+
+#include <list>
+#include <tuple>
+
+namespace slate {
+
+namespace impl {
+
+//------------------------------------------------------------------------------
+/// @internal
+/// Distributed parallel general matrix 2-norm estimate.
+/// Generic implementation for any target.
+/// @ingroup norm_specialization
+///
+template <Target target, typename matrix_type>
+blas::real_type<typename matrix_type::value_type>
+norm2est(
+    matrix_type A,
+    Options const& opts)
+{
+    using scalar_t = typename matrix_type::value_type;
+    using real_t = blas::real_type<scalar_t>;
+    using blas::min;
+
+    int p, q;
+    int myrow, mycol;
+    int izero = 0;
+    GridOrder order;
+    A.gridinfo(&order, &p, &q, &myrow, &mycol);
+
+    int64_t n = A.n();
+    int64_t m = A.m();
+    int64_t mb = A.tileMb(0);
+    int64_t nb = A.tileNb(0);
+
+    scalar_t one = 1.;
+    scalar_t zero  = 0.;
+    real_t rone = 1.;
+    real_t alpha;
+
+    int64_t cnt = 0;
+    int64_t maxiter = min( 100, n );
+
+    real_t e  = 0.;
+    real_t e0  = 0.;
+    real_t normX = 0;
+    real_t normAX = 0;
+    real_t tol = 1.e-1;
+
+    int64_t mloc   = numberLocalRowOrCol(m, mb, myrow, izero, p);
+    int64_t nloc   = numberLocalRowOrCol(n, nb, mycol, izero, q);
+    int64_t lldA   = blas::max(1, mloc);
+    int64_t lldA_n = blas::max(1, nloc);
+
+    // todo: do we still need reserveDevice here?
+    if (target == Target::Devices)
+        A.reserveDeviceWorkspace();
+
+    std::vector<real_t> local_sums(n);
+    std::vector<scalar_t> global_sums(n);
+    std::vector<scalar_t> W1(lldA);
+
+    auto XL = slate::Matrix<scalar_t>::fromScaLAPACK(
+            n, 1, &global_sums[0], n, nb, 1, p, q, A.mpiComm());
+    auto AX = slate::Matrix<scalar_t>::fromScaLAPACK(
+            m, 1, &W1[0], lldA, mb, 1, p, q, A.mpiComm());
+
+    // Two norm estimation
+    // First: let's compute the x vector such that
+    // x_j = sum_i abs( A_{i,j} ), x here is global_sums.data
+    #pragma omp parallel
+    #pragma omp master
+    {
+        internal::norm<target>(slate::Norm::One, NormScope::Matrix, std::move(A), local_sums.data());
+    }
+
+    // global_sums.data() will have the sum of each column.
+    #pragma omp critical(slate_mpi)
+    {
+        trace::Block trace_block("MPI_Allreduce");
+        slate_mpi_call(
+                MPI_Allreduce(local_sums.data(), global_sums.data(),
+                    n, mpi_type<real_t>::value,
+                    MPI_SUM, A.mpiComm()));
+    }
+
+    // Second: compute the ||x||_Fro
+    e  = slate::norm(Norm::Fro, XL, opts);
+    if (e == 0.) {
+        return 0.;
+    }
+    else {
+        normX = e;
+    }
+
+    // Third: start the while-loop X = X / ||X||
+    while ((cnt < maxiter) && (fabs((e - e0)) > (tol * (e)))) {
```

Comment:
Prefer using `std::abs`. There are an excess of parens here. Why does `fabs((e - e0))` need 2, or `(e)` need any?

Reply:
Sure, changed on the stat and forgot these.